openebs/pkg/volume/provision_v1alpha1.go

```diff
@@ -0,0 +1,201 @@
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *openEBSProvisionerV1alpha1) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+
+	//Issue a request to Maya API Server to create a volume
+	var volume mayav1.Volume
+	var openebsCASVol mApiv1.OpenEBSVolumeV1Alpha1
+	casVolume := v1alpha1.CASVolume{}
+
+	volSize := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	casVolume.Spec.Capacity = volSize.String()
+
+	className := GetStorageClassName(options)
+
+	// creating a map b/c have to initialize the map using the make function before
+	// adding any elements to avoid nil map assignment error
+	mapLabels := make(map[string]string)
+
+	if className == nil {
+		glog.Errorf("Volume has no storage class specified")
+	} else {
+		mapLabels[string(v1alpha1.StorageClassKey)] = *className
+		glog.Infof("Using Storageclass: %s", *className)
+		casVolume.Labels = mapLabels
+	}
+	casVolume.Labels[string(v1alpha1.NamespaceKey)] = options.PVC.Namespace
+	casVolume.Namespace = options.PVC.Namespace
+	casVolume.Labels[string(v1alpha1.PersistentVolumeClaimKey)] = options.PVC.ObjectMeta.Name
+	casVolume.Name = options.PVName
+
+	//Pass through labels from PVC to maya-apiserver
+	//volumeSpec.Metadata.Labels.Application = options.PVC.ObjectMeta.GetLabels()[mayav1.PVCLabelsApplication]
+	//volumeSpec.Metadata.Labels.ReplicaTopoKeyDomain = options.PVC.ObjectMeta.GetLabels()[mayav1.PVCLabelsReplicaTopKeyDomain]
+	//volumeSpec.Metadata.Labels.ReplicaTopoKeyType = options.PVC.ObjectMeta.GetLabels()[mayav1.PVCLabelsReplicaTopKeyType]
+
+	_, err := openebsCASVol.CreateVolume(casVolume)
+	if err != nil {
+		glog.Errorf("Error creating volume: %v", err)
+		return nil, err
+	}
+
+	err = openebsCASVol.ReadVolume(options.PVName, options.PVC.Namespace, &volume)
+	if err != nil {
+		glog.Errorf("Error getting volume details: %v", err)
+		return nil, err
+	}
+
+	// Use annotations to specify the context using which the PV was created.
+	volAnnotations := make(map[string]string)
+	volAnnotations["openEBSProvisionerIdentity"] = p.identity
+
+	var iqn, targetPortal string
+
+	for key, value := range volume.Metadata.Annotations.(map[string]interface{}) {
+		switch key {
+		case "vsm.openebs.io/iqn":
+			iqn = value.(string)
+		case "vsm.openebs.io/targetportals":
+			targetPortal = value.(string)
+		}
+	}
+
+	glog.V(2).Infof("Volume IQN: %v , Volume Target: %v", iqn, targetPortal)
+
+	if !util.AccessModesContainedInAll(p.GetAccessModes(), options.PVC.Spec.AccessModes) {
+		glog.V(1).Info("Invalid Access Modes: %v, Supported Access Modes: %v", options.PVC.Spec.AccessModes, p.GetAccessModes())
+		return nil, fmt.Errorf("Invalid Access Modes: %v, Supported Access Modes: %v", options.PVC.Spec.AccessModes, p.GetAccessModes())
+	}
+
+	// The following will be used by the dashboard, to display links on PV page
+	userLinks := make([]string, 0)
```

Comment:
Can we put these logic somewhere else ? modularize?

Reply:
done.